File: examples/streaming_client.py
"""Minimal streaming inference client.

Subscribes to continuous predictions for an entity and prints each response
as it arrives. Runs until Ctrl-C.

Usage:
    pip install grpcio grpcio-tools
    python -m grpc_tools.protoc \
        -I proto \
        --python_out=. \
        --grpc_python_out=. \
        proto/axon/inference/v1/inference.proto

    python examples/streaming_client.py
    python examples/streaming_client.py localhost:50052
"""

import sys

import grpc
from axon.inference.v1 import inference_pb2 as pb
from axon.inference.v1 import inference_pb2_grpc as stub

addr = sys.argv[1] if len(sys.argv) > 1 else "localhost:50051"

try:
    with grpc.insecure_channel(addr) as channel:
        client = stub.InferenceServiceStub(channel)
        for response in client.PredictStream(
            pb.PredictStreamRequest(entity_id="example-entity")
        ):
            for output in response.outputs:
                print(f"[{response.timestamp_ms}] {output.name}: {list(output.values)}")
except KeyboardInterrupt:
    pass
except grpc.RpcError as e:
    print(f"error: {e.details()}")
    print("hint: is the server running?  try: axon serve --config config.toml")
    sys.exit(1)

File: examples/streaming_client.rs
//! Streaming inference client.
//!
//! Subscribes to continuous predictions for a single entity. Prints each
//! prediction as it arrives. The server pushes a new prediction each time the
//! entity's features update in the store.
//!
//! Press Ctrl-C to stop.
//!
//! # Usage
//!
//! ```bash
//! # against a local server on the default port
//! cargo run --example streaming_client
//!
//! # against a specific address
//! cargo run --example streaming_client -- http://[::1]:50051
//! ```

use anyhow::{Context, Result};

mod proto {
    tonic::include_proto!("axon.inference.v1");
}

use proto::PredictStreamRequest;
use proto::inference_service_client::InferenceServiceClient;

#[tokio::main]
async fn main() -> Result<()> {
    let addr = std::env::args()
        .nth(1)
        .unwrap_or_else(|| "http://[::1]:50051".to_owned());

    let mut client = InferenceServiceClient::connect(addr.clone())
        .await
        .with_context(|| {
            format!(
                "could not connect to axon at '{addr}'\n  hint: is the server running?\
            \n        try: axon serve --config config.toml"
            )
        })?;

    let request = tonic::Request::new(PredictStreamRequest {
        entity_id: "example-entity".to_owned(),
        metadata: Default::default(),
    });

    let mut stream = client
        .predict_stream(request)
        .await
        .context("predict_stream RPC failed")?
        .into_inner();

    println!("streaming predictions for 'example-entity' — press Ctrl-C to stop");
    println!();

    loop {
        tokio::select! {
            msg = stream.message() => {
                match msg.context("stream error")? {
                    None => {
                        println!("stream closed by server");
                        break;
                    }
                    Some(resp) => {
                        print!("[{}ms] ", resp.timestamp_ms);
                        for output in &resp.outputs {
                            print!("{}={:?}  ", output.name, output.values);
                        }
                        println!();
                    }
                }
            }
            _ = tokio::signal::ctrl_c() => {
                println!("\nshutdown signal received, stopping");
                break;
            }
        }
    }

    Ok(())
}
